standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/RetryingHMSHandler.java

```diff
@@ -203,11 +203,12 @@ public Result invokeInternal(final Object proxy, final Method method, final Obje
         }
       }
 
-      if (retryCount >= retryLimit) {
+      Throwable rootCause = ExceptionUtils.getRootCause(caughtException);
+      String errorMessage = ExceptionUtils.getMessage(caughtException) +
```

Comment:
In this case, the `caughtException` is always `JDOException`. We should be able to get the deep cause and should not have the duplication message. Also I think the root cause can help users understand why this exception happens.

Reply:
Users are not necessarily developers; throwing exceptions back to them should be the norm. Consider for instance some of the most popular DBMS (Postgres, Oracle, MSSQL, etc); I don't think anyone is exposing exceptions to the user.

This is a minor comment not really blocking but I wanted to mention that propagating exceptions to the client is not the ideal solution.

Reply:
In the prod environment, HMS may be transparent to users. And since we will pass the exception messages to the client, why not pass the most useful messages?

Reply:
This is an improvement given the current situation and that's why I think its worth merging. My comments are more forward looking on things that we could improve in the future.